src/nbl/ui/CWindowWin32.cpp

```diff
@@ -1,283 +1,375 @@
-			// Using ToUnicode instead would be shorter, but would to my knowledge not run on 95 and 98.
-			WORD keyChars[2];
-			UINT scanCode = HIWORD(lParam);
-			int conversionResult = ToAsciiEx(wParam, scanCode, allKeys, keyChars, 0, KEYBOARD_INPUT_HKL);
-			if (conversionResult == 1)
+		void background_work()
+		{
+			static MSG message;
+			static uint32_t timeoutInMS = 8; // gonna become 10 anyway
+			if (getMessageWithTimeout(&message, timeoutInMS))
 			{
-				WORD unicodeChar;
-				MultiByteToWideChar(
-					KEYBOARD_INPUT_CODEPAGE,
-					MB_PRECOMPOSED, // default
-					(LPCSTR)keyChars,
-					sizeof(keyChars),
-					(WCHAR*)&unicodeChar,
-					1);
-				event.KeyInput.Char = unicodeChar;
+				TranslateMessage(&message);
+				DispatchMessage(&message);
 			}
-			else
-				event.KeyInput.Char = 0;
-
-			// allow composing characters like '@' with Alt Gr on non-US keyboards
-			if ((allKeys[VK_MENU] & 0x80) != 0)
-				event.KeyInput.Control = 0;
-
-			dev = getDeviceFromHWnd(hWnd);
-			if (dev)
-				dev->postEventFromUser(event);
+		}
 
-			if (message == WM_SYSKEYDOWN || message == WM_SYSKEYUP)
-				return DefWindowProc(hWnd, message, wParam, lParam);
-			else
-				return 0;
+		void process_request(SRequest& req)
+		{
+			req.process();
 		}
 
-		case WM_SIZE:
+		template <typename RequestParams>
+		void request_impl(SRequest& req, RequestParams&& params)
+		{
+			req.setParams(std::move(params));
+		}
+	private:
+		static bool getMessageWithTimeout(MSG* msg, uint32_t timeoutInMilliseconds)
 		{
-			// resize
-			dev = getDeviceFromHWnd(hWnd);
-			if (dev)
-				dev->OnResized();
+			bool res;
+			UINT_PTR timerId = SetTimer(NULL, NULL, timeoutInMilliseconds, NULL);
+			res = GetMessage(msg, nullptr, 0, 0);
+			
+			PostMessage(nullptr, WM_NULL, 0, 0);
+			//KillTimer(NULL, timerId);
+
+			if (!res)
+				return false;
+			if (msg->message == WM_TIMER && msg->hwnd == NULL && msg->wParam == timerId)
+				return false;
+			return true;
 		}
-		return 0;
+	} windowThreadHandler; 
 
-		case WM_DESTROY:
-			PostQuitMessage(0);
-			return 0;
+	CWindowWin32::CWindowWin32(core::smart_refctd_ptr<system::ISystem>&& sys, int _x, int _y, uint32_t _w, uint32_t _h, E_CREATE_FLAGS _flags)
+	{
+		m_width = _w; m_height = _h;
+		windowThreadHandler.createWindow(_x, _y, _w, _h, _flags, m_native);
+		SetWindowLongPtr(m_native, GWLP_USERDATA, reinterpret_cast<LONG_PTR>(this));
+	}
 
-		case WM_SYSCOMMAND:
-			// prevent screensaver or monitor powersave mode from starting
-			if ((wParam & 0xFFF0) == SC_SCREENSAVE ||
-				(wParam & 0xFFF0) == SC_MONITORPOWER ||
-				(wParam & 0xFFF0) == SC_KEYMENU
-				)
-				return 0;
+	CWindowWin32::~CWindowWin32()
+	{
+		windowThreadHandler.destroyWindow(m_native);
+	}
 
+    LRESULT CALLBACK CWindowWin32::WndProc(HWND hWnd, UINT message, WPARAM wParam, LPARAM lParam)
+	{
+		CWindowWin32* window = reinterpret_cast<CWindowWin32*>(GetWindowLongPtr(hWnd, GWLP_USERDATA));
```

Comment:
oh, i see thats how you resolved HWND->CWindowWin32 mapping, looks neat
can you point me where you're setting this user data? out of curiosity

Reply:
Yeah in the CWindowWin32 ctor